System/Random.hs

```diff
@@ -20,41 +20,181 @@
+--     how to write a 'RandomGen' instance.
+--
+-- *   This library provides instances for 'Random' for some unbounded datatypes
+--     for backwards compatibility. For an unbounded datatype, there is no way to
+--     generate a value with uniform probability out of its entire domain, so the
+--     'random' implementation for unbounded datatypes actually generates a value
+--     based on some fixed range.
+--
+--     For 'Integer', 'random' generates a value in the 'Int' range. For 'Float' and 'Double', 'random' generates a floating point value in the range @[0, 1)@.
+--
+--     This library does not provide 'Uniform' instances for any unbounded datatypes.
+--
+-- = Reproducibility
+--
+-- If you have two builds of a particular piece of code against this library,
+-- any deterministic function call should give the same result in the two
+-- builds if the builds are
+--
+-- *   compiled against the same major version of this library
+-- *   on the same architecture (32-bit or 64-bit)
+--
+-- = Pure and monadic pseudo-random number generators
+--
+-- Pseudo-random number generators come in two flavours: /pure/ and /monadic/.
+--
+-- Pure pseudo-random number generators produce a new random value together
+-- with a new instance of the pseudo-random number generator. 'RandomGen' defines
+-- the interface for pure pseudo-random number generators.
+--
+-- Monadic pseudo-random number generators mutate their own state as they
+-- produce random values. They generally live in 'ST' or 'IO' or some transformer
+-- that implements 'PrimMonad'. 'MonadRandom' defines the interface for monadic
+-- pseudo-random number generators.
+--
+-- Pure pseudo-random number generators can be used in monadic code via the
+-- adapters 'PureGen', 'PrimGen' and 'MutGen'.
+--
+-- *   'PureGen' can be used in any state monad. With strict 'StateT' there is
+--     even no performance overhead when compared to the 'RandomGen' directly.
+--     But it is not safe to use it in presence of exceptions and resource
+--     allocation that requires cleanup.
+--
+-- *   'PrimGen' can be used in any 'PrimMonad' if the pseudo-random number
+--     generator is also an instance of 'Prim'. It will perform much faster
+--     than 'MutGen', but it can't be used in a concurrent setting.
+--
+-- *   'MutGen' can be used in any 'PrimMonad' (this includes 'ST' and 'IO') and
+--     can safely be shared between threads.
+--
+-- When to use which?
+--
+-- *   Use 'PureGen' if at the end your computation will not have side effects
+--     and will result in a pure function.
+--
+-- *   Use 'PrimGen' if the pseudo-random number generator implements 'Prim'
+--     class and random value generation is intermixed with 'IO' or 'ST'.
+--
+-- *   Otherwise use 'MutGen'. Whenever a 'MutGen' is shared between threads,
+--     make sure there is not much contention for it, otherwise performance
+--     will be horrendous. For parallel random value generation it is best to
```

Comment:
missing "to"?

Reply:
Interesting. I assume you're referring to "use" vs. "to use".

I'm parsing this sentence as follows:

{{ For parallel random value generation }} it is best to {{ split the generator and [to] use a single 'PureGen' or 'PrimGen' per thread }}.

Where {{ ... }} indicates grammatical grouping. In this first parse tree, it feels to me like no "to" is required before the word "use". However, the situation is different when the sentence is parsed as follows:

{{ For parallel random value generation }} it is best {{ to split the generator }} and {{ [to] use a single 'PureGen' or 'PrimGen' per thread }}.

Here it feels more natural to have the "to".

Reply:
Ok it seems fine - I think I read it in the code and missed the preceeding bit of context